cylc/flow/config.py

```diff
@@ -1008,6 +1016,218 @@ def _check_sequence_bounds(self):
+                ' in completion expressions.'
+            )
+        elif '-' in expr:
+            raise WorkflowConfigError(
+                f'Error in [runtime][{task_name}]completion:'
+                f'\n  {expr}'
+                '\nReplace hyphens with underscores in task outputs when'
+                ' used in completion expressions.'
+            )
+
+        # get the outputs and completion expression for this task
+        try:
+            outputs = self.taskdefs[task_name].outputs
+        except KeyError:
+            # this is a family -> we'll check integrity for each task that
+            # inherits from it
+            return
+
+        (
+            trigger_to_completion_variable,
+            completion_variable_to_trigger,
+        ) = get_trigger_completion_variable_maps(outputs.keys())
+
+        # get the optional/required outputs defined in the graph
+        graph_optionals = {
+            # completion_variable: is_optional
+            trigger_to_completion_variable[trigger]: (
+                None if is_required is None else not is_required
+            )
+            for trigger, (_, is_required)
+            in outputs.items()
+        }
+        if (
+            graph_optionals[TASK_OUTPUT_SUCCEEDED] is True
+            and graph_optionals[TASK_OUTPUT_FAILED] is None
+        ):
+            # failed is implicitly optional if succeeded is optional
+            # https://github.com/cylc/cylc-flow/pull/6046#issuecomment-2059266086
+            graph_optionals[TASK_OUTPUT_FAILED] = True
+
+        # get the optional/required outputs defined in the expression
+        try:
+            # this involves running the expression which also validates it
+            expression_optionals = get_optional_outputs(expr, outputs)
+        except NameError as exc:
+            # expression references an output which has not been registered
+            error = exc.args[0][5:]
+
+            if f"'{TASK_OUTPUT_FINISHED}'" in error:
+                # the finished output cannot be used in completion expressions
+                # see proposal point 5::
+                # https://cylc.github.io/cylc-admin/proposal-optional-output-extension.html#proposal
+                raise WorkflowConfigError(
+                    f'Error in [runtime][{task_name}]completion:'
+                    f'\n  {expr}'
+                    '\nThe "finished" output cannot be used in completion'
+                    ' expressions, use "succeeded or failed".'
+                )
+
+            for alt_qualifier, qualifier in ALT_QUALIFIERS.items():
+                if alt_qualifier in error:
+                    raise WorkflowConfigError(
+                        f'Error in [runtime][{task_name}]completion:'
+                        f'\n  {expr}'
+                        f'\nUse "{qualifier}" not "{alt_qualifier}" '
+                        'in completion expressions.'
+                    )
```

Comment:
Not quite getting the right result from this
```
WorkflowConfigError: Error in [runtime][a]completion:
  succeeded and (x or submit_fail)
Use "submitted" not "submit" in completion expressions.
```
I think it could be moved to after here:
https://github.com/cylc/cylc-flow/blob/23567bc991111f9e178c3d13ca4112a5deb0ac2f/cylc/flow/task_outputs.py#L230-L231

Reply:
Resolved this via a different approach to keep this in the validate code (also needed to convert qualifier to compvar).